### PR TITLE
DF-209: Split datefield (componentised)

### PR DESCRIPTION
**Pull request by ababic**
Supersedes #385

### CODE REVIEW

File: etna/core/tests/test_fields.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    def test_end_date_default_day31_month12_for_empty_day_empty_month(self):
+        input_day = ""
+        input_month = ""
+        input_year = "2001"
+        date = datetime.date(year=int(input_year), month=12, day=31)
+        field = DateInputField(
+            required=False, default_day=END_OF_MONTH, default_month=12
+        )
+        value = field.clean([input_day, input_month, input_year])
+        self.assertEqual(value, date)
```
Comment (ababic) [inline etna/core/tests/test_fields.py:85]: Suggestions:
- Avoid unnecessary variabalisation: If a value is only used once, variabalising it makes the code more difficult to scan
- Try to limit tests to 'one thing' per test, or use docstrings to explain what is being tested where that is not possible
- Where there is a pattern to what you are testing, loop over variations of inputs / expected outputs and use `subTest`... there's no better way to indicate "this same thing is being tested each time"

File: etna/search/tests/test_forms.py [hunk 2]
```diff
@@ -4,12 +4,89 @@
...
-    def test_start_date_after_end_date_is_invalid(self):
+    def test_empty_opening_start_date_is_valid(self):
         form = CatalogueSearchForm(
             {
                 "group": "tna",
-                "opening_start_date": "2000-01-01",
-                "opening_end_date": "1900-01-01",
+                "opening_end_date_0": "01",
+                "opening_end_date_1": "12",
+                "opening_end_date_2": "2002",
```
Comment (ababic) [inline etna/search/tests/test_forms.py:13]:
```suggestion
                "opening_start_date_0": "01",
                "opening_start_date_1": "12",
                "opening_start_date_2": "2002",
```

File: etna/search/tests/test_forms.py [hunk 3]
```diff
@@ -4,12 +4,89 @@
...
+
+        self.assertTrue(is_valid)
+
+    def test_empty_opening_end_date_is_valid(self):
+        form = CatalogueSearchForm(
+            {
+                "group": "tna",
+                "opening_start_date_0": "01",
+                "opening_start_date_1": "01",
+                "opening_start_date_2": "2001",
```
Comment (ababic) [inline etna/search/tests/test_forms.py:27]:
```suggestion
                "opening_end_date_0": "01",
                "opening_end_date_1": "01",
                "opening_end_date_2": "2001",
```

File: etna/core/tests/test_fields.py (hunk 1 above)
Comment (TNA-Allan) [reply etna/core/tests/test_fields.py]: Thanks @ababic updated.

File: etna/search/tests/test_forms.py (hunk 2 above)
Comment (TNA-Allan) [reply etna/search/tests/test_forms.py]: The case was correct, but Not relevant, code changed to subTest.

File: etna/search/tests/test_forms.py [hunk 4]
```diff
@@ -4,12 +4,89 @@
...
+                "opening_start_date_1": "12",
+                "opening_start_date_2": "1999",
+                "opening_end_date_1": "01",
+                "opening_end_date_2": "2000",
+            }
+        )
+
+        is_valid = form.is_valid()
+
+        self.assertTrue(is_valid)
```
[the remark this answers is not in the document]
Comment (TNA-Allan) [reply etna/search/tests/test_forms.py]: @ababic , Thanks for this

> These tests are overkill.
Had to retain some due to its nature of using both fields during validation. For example in `test_various_opening_date_inputs_is_valid`
Others moved to `test_fields.py` and added `test_validators.py`

File: etna/core/tests/test_fields.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                "Entered date must include a Month.",
+            ),
+            (
+                "input year is empty ddmm<empty>",
+                {"input_day": "01", "input_month": "01", "input_year": ""},
+                "Entered date must include a Year.",
+            ),
+        ):
+            with self.subTest(label):
+                field = DateInputField(
```
Comment (ababic) [inline etna/core/tests/test_fields.py:100]: Field could be defined outside of the loop, but it's fine here

File: etna/core/tests/test_fields.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                "Entered date must be a real date, for example 23 9 2017.",
+            ),
+            (
+                "input day is empty <empty>mmyyyy",
+                {"input_day": "", "input_month": "01", "input_year": "2007"},
+                "Entered date must be a real date, for example 23 9 2017.",
+            ),
+        ):
+            with self.subTest(label):
+                field = DateInputField(required=False)
```
Comment (ababic) [inline etna/core/tests/test_fields.py:128]: Field could be defined outside of the loop, but it's fine here

File: etna/core/tests/test_validators.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
+from django.core.exceptions import ValidationError
+from django.test import SimpleTestCase
+
+from etna.core.fields import DateInputField
+
+
+class DateFieldValidatorTest(SimpleTestCase):
```
Comment (ababic) [inline etna/core/tests/test_validators.py:7]: From the name of the module, I would expect this file to contain unit tests for the validators themselves, but we are testing the field again here. I would just move this test case to `test_fields.py` and remove this module.

File: etna/search/tests/test_forms.py [hunk 8]
```diff
@@ -4,31 +4,79 @@
...
 
 class CatalogueSearchFormTest(SimpleTestCase):
-    def test_start_date_after_end_date_is_invalid(self):
+    def test_various_opening_date_inputs_is_valid(self):
+        """tests inputs accross both date fields with various input formats for form validity"""
+        for label, value in (
+            (
+                "input end date with empty start date",
+                {
+                    "input": {
```
Comment (ababic) [inline etna/search/tests/test_forms.py:13]: Not sure what we're gaining from the multi-level dicts here. If all we need is data for the form, we should lose the outer dicts (with the `"input"` key)

File: etna/search/tests/test_forms.py [hunk 9]
```diff
@@ -4,31 +4,79 @@
 
 
 class CatalogueSearchFormTest(SimpleTestCase):
-    def test_start_date_after_end_date_is_invalid(self):
+    def test_various_opening_date_inputs_is_valid(self):
+        """tests inputs accross both date fields with various input formats for form validity"""
+        for label, value in (
```
Comment (ababic) [inline etna/search/tests/test_forms.py:9]:
```suggestion
        for label, form_data in (
```

File: etna/search/tests/test_forms.py [hunk 10]
```diff
@@ -4,31 +4,79 @@
...
+                        "opening_start_date_2": "1999",
+                        "opening_end_date_0": "31",
+                        "opening_end_date_1": "12",
+                        "opening_end_date_2": "1999",
+                    }
+                },
+            ),
+        ):
+            with self.subTest(label):
+                form = CatalogueSearchForm(value["input"])
```
Comment (ababic) [inline etna/search/tests/test_forms.py:62]:
```suggestion
                form = CatalogueSearchForm(form_data)
```

File: etna/core/fields.py [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+                required=default_day is None,
+            ),
+            forms.CharField(
+                label=_("Month"),
+                error_messages={"incomplete": "Entered date must include a Month."},
+                validators=[
+                    PositiveIntegerStringValidator(
+                        min=1,
+                        max=12,
+                        msg="Entered date must be a real date, for example 23 9 2017.",
```
Comment (ababic) [inline etna/core/fields.py:67]: These are component-specific errors, so making them all generic makes the field less helpful. What was wrong with the original wording?

Comment (ababic) [reply etna/core/fields.py:67]: There's definitely no reason the 'Day', 'Month' or 'Year' to be capitalised in any of these messages

File: etna/core/fields.py [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+                field_value = None
+            if field_value in self.empty_values:
+                if self.required and self.require_all_fields:
+                    # Raise a 'required' error if the MultiValueField is
+                    # required and all sub-fields are required.
+                    raise ValidationError(
+                        self.error_messages["required"], code="required"
+                    )
+
+                if (
```
Comment (ababic) [inline etna/core/fields.py:143]: This new condition prevents a default month value from being respected below

File: etna/core/fields.py [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+                    i == 1
+                    and self.default_values[i]
+                    and value[0]
+                    and value[2]
+                    and not value[1]
+                ):
+                    # Validate empty month input (overrid default month) when day, year are input
+                    msg = "Entered date must include a Month."
```
Comment (ababic) [inline etna/core/fields.py:151]: Why are we giving month special treatment here? If no default is specified for month, `field.required` will be `True`, and a widget-specific incomplete error will be raised below.

File: etna/core/fields.py [hunk 14]
```diff
@@ -153,14 +187,24 @@ def clean(self, value):
...
+            raise ValidationError(e)
 
         try:
             out = self.compress(clean_data)
         except ValueError:
-            # The specified day must be invalid for the month, so add an error message
-            # to the 'day' sub-field widget before raising
-            msg = "Enter a valid day for the month you entered"
+            # When Day, Month belong to correct range and Year is 0, then error-value "year 0 is out of range"
+            msg = "Entered date must be a real date, for example 23 9 2017."
```
Comment (ababic) [inline etna/core/fields.py:209]: What's the benefit of making this generic? Surely that just makes it difficult to see what is wrong?

File: etna/core/tests/test_fields.py (hunk 5 above)
Comment (TNA-Allan) [reply etna/core/tests/test_fields.py]: Done - its outside the loop.

File: etna/core/tests/test_fields.py (hunk 6 above)
Comment (TNA-Allan) [reply etna/core/tests/test_fields.py]: Done - its outside the loop.

File: etna/core/tests/test_validators.py (hunk 7 above)
Comment (TNA-Allan) [reply etna/core/tests/test_validators.py:7]: As mentioned in the doc string, it validates the overridden default messages. I've split the test case for each field validator.

File: etna/search/tests/test_forms.py (hunk 8 above)
Comment (TNA-Allan) [reply etna/search/tests/test_forms.py]: Done - removed input key

File: etna/search/tests/test_forms.py (hunk 9 above)
Comment (TNA-Allan) [reply etna/search/tests/test_forms.py]: Same as above Done - removed input key

File: etna/search/tests/test_forms.py (hunk 10 above)
Comment (TNA-Allan) [reply etna/search/tests/test_forms.py]: Same as above Done - removed input key

File: etna/core/fields.py (hunk 11 above)
Comment (TNA-Allan) [reply etna/core/fields.py:67]: This the requirement.

File: etna/core/fields.py (hunk 12 above)
Comment (TNA-Allan) [reply etna/core/fields.py:143]: It does respect default month  see 
 `etna.core.tests.test_fields.DateFieldTest.test_various_inputs_with_dtf_default_for_end_date`

File: etna/core/fields.py (hunk 14 above)
Comment (TNA-Allan) [reply etna/core/fields.py:209]: This is the requirement for those error handling.

**Review by TNA-Allan — APPROVED** [review]
Looks good :)

**Pull request merged**